src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs

```diff
@@ -0,0 +1,136 @@
+            {
+                throw new ArgumentNullException(nameof(errorReporter));
+            }
+
+            _documentFactory = documentFactory;
+            _errorReporter = errorReporter;
+        }
+
+        public override object GetHostProject(ITextBuffer textBuffer)
+        {
+            if (textBuffer == null)
+            {
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            // If there's no document we can't find the FileName, or look for an associated project.
+            if (!_documentFactory.TryGetTextDocument(textBuffer, out var textDocument))
+            {
+                return null;
+            }
+
+            var projectsContainingFilePath = IdeApp.Workspace.GetProjectsContainingFile(textDocument.FilePath);
+            foreach (var project in projectsContainingFilePath)
+            {
+                if (!(project is DotNetProject))
+                {
+                    continue;
+                }
+
+                var projectFile = project.GetProjectFile(textDocument.FilePath);
+                if (!projectFile.IsHidden)
+                {
+                    return project;
+                }
+            }
+
+            return null;
+        }
+
+        public override string GetProjectPath(object project)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var dotnetProject = project as DotNetProject;
+
+            Debug.Assert(dotnetProject != null);
+
+            return dotnetProject.FileName.FullPath;
+        }
+
+        public override bool IsSupportedProject(object project)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var dotnetProject = (DotNetProject)project;
+
+            Debug.Assert(dotnetProject != null);
+
+            try
+            {
+                // ASP.NET Core Web projects
+                if (dotnetProject.IsCapabilityMatch(AspNetCoreWebCapability))
+                {
+                    return true;
+                }
+
+                // ASP.NET Core Class libraries
+                if (dotnetProject.References.Any(reference => reference.Reference.StartsWith(MvcAssemblyName, StringComparison.Ordinal)))
+                {
+                    return true;
+                }
+            }
+            catch (Exception ex)
+            {
+                _errorReporter.ReportError(ex);
+            }
+
+            return false;
+        }
+
+        public override string GetProjectName(object project)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var dotnetProject = (DotNetProject)project;
+
+            Debug.Assert(dotnetProject != null);
+
+            return dotnetProject.Name;
```

Comment:
I think you may have been on something when you wrote these asserts.

Reply:
lol looking back at these totally agree. Good submission to shitcode 😉